pkg/fetcher/fetcher.go

```diff
@@ -414,6 +428,56 @@ func (u *URLFetcher) IsNonretryableClientError() bool {
 		s != http.StatusLocked && s != http.StatusTooManyRequests
 }
 
+// buildRegistryErrMsg builds error message for unexpected http code (nonretryable client errors and all other errors)
+// and extracts message details from response body stream if there is one (#5951).
+func (u *URLFetcher) buildRegistryErrMsg(httpStatusCode int, url *url.URL, respBody io.ReadCloser) string {
+	errMsg := fmt.Sprintf("Unexpected http code: %d (%s), URL: %s", httpStatusCode, http.StatusText(httpStatusCode), url)
+
+	errDetail, err := u.extractErrResponseMessage(respBody)
+	if err != nil {
+		return errMsg
+	}
+
+	errMsg += fmt.Sprintf(", Message: %s", errDetail)
+	return errMsg
+}
+
+// extractErrResponseMessage extracts `message` field from error response body json (#5951).
+func (u *URLFetcher) extractErrResponseMessage(rdr io.ReadCloser) (string, error) {
```

Comment:
Bonus: doing this would make it easier to unit test this function, which we should. You can write a unit test in a new file, `pkg/fetcher/fetcher_test.go`.